packages/debugger/src/panels/kernelSources/body.ts

```diff
@@ -0,0 +1,215 @@
+// Copyright (c) Jupyter Development Team.
+// Distributed under the terms of the Modified BSD License.
+
+import {
+  CodeEditorWrapper,
+  IEditorMimeTypeService,
+  IEditorServices
+} from '@jupyterlab/codeeditor';
+
+import { viewBreakpointIcon } from '../../icons';
+
+import { ToolbarButton } from '@jupyterlab/ui-components';
+
+import { Kernel, KernelMessage } from '@jupyterlab/services';
+
+import { Signal } from '@lumino/signaling';
+
+import { PanelLayout, Widget } from '@lumino/widgets';
+
+import { Debugger } from '../..';
+
+import { EditorHandler } from '../../handlers/editor';
+
+import { IDebugger } from '../../tokens';
+
+/**
+ * The body for a Sources Panel.
+ */
+export class KernelSourcesBody extends Widget {
+  /**
+   * Instantiate a new Body for the SourcesBody widget.
+   *
+   * @param options The instantiation options for a SourcesBody.
+   */
+  constructor(options: KernelSourcesBody.IOptions) {
+    super();
+    this._model = options.model;
+    this._debuggerService = options.service;
+    this._mimeTypeService = options.editorServices.mimeTypeService;
+
+    const factory = new Debugger.ReadOnlyEditorFactory({
+      editorServices: options.editorServices
+    });
+
+    const layout = new PanelLayout();
+    this.layout = layout;
+    this.addClass('jp-DebuggerKernelSources-body');
+
+    this._editor = factory.createNewEditor({
+      content: '',
+      mimeType: '',
+      path: ''
+    });
+    this._editor.hide();
+
+    this._model.currentFrameChanged.connect(async (_, frame) => {
+      if (!frame) {
+        this._clearEditor();
+        return;
+      }
+      const kernel = this._debuggerService.session?.connection
+        ?.kernel as Kernel.IKernelConnection;
+      kernel.registerCommTarget(
+        'get_sources',
+        (comm: Kernel.IComm, commOpen: KernelMessage.ICommOpenMsg) => {
+          comm.onMsg = msg => {
+            const data = JSON.parse(msg.content.data as any);
+            Object.entries(data).forEach(entry => {
+              const [key, value] = entry;
+              const button = new ToolbarButton({
+                icon: viewBreakpointIcon,
+                onClick: (): void => {
+                  this._debuggerService
+                    .getSource({
+                      sourceReference: 0,
+                      path: value as string
+                    })
+                    .then(source => {
+                      console.log('---', source);
+                      this._model.currentSource = source;
+                      this._model.open();
+                    });
+                },
+                label: key,
+                tooltip: value as string
+              });
+              layout.addWidget(button);
+            });
+          };
+        }
+      );
+      const code = `
+import sys, re, json
```

Comment:
Since this is Python specific, is there another language agnostic way for retrieving the sources?

Reply:
Or is this mostly a workaround for `supportsLoadedSourcesRequest` as mentioned in https://github.com/jupyterlab/jupyterlab/issues/10139#issuecomment-872169353?

Reply:
It is indeed a workaround on the lack of support for `SourcesRequest`. I have tried to use Requests_Modules https://microsoft.github.io/debug-adapter-protocol/specification#Requests_Modules but it returns only a partial list of the expected modules, not sure why (in my case 13 modules, missing a lot of the existing and loaded modules).

I guess the Kernel sources could be activated in first instance for python kernels only. Not sure how to identify the kernel is python...?

Reply:
There is `language_info` as part of the kernel info reply. Checking that and whether debugging is available could be enough to activate the kernel sources.

Another way would be to improve the extensibility of the core jupyterlab debugger / sidebar so it's easy to add new panels, if that's not the case already. And implement the Python specific kernel sources logic in a separate extension.

Reply:
> There is language_info as part of the kernel info reply. Checking that and whether debugging is available could be enough to activate the kernel sources.

Thx, will implement that in this PR.

> Another way would be to improve the extensibility of the core jupyterlab debugger / sidebar so it's easy to add new panels, if that's not the case already. And implement the Python specific kernel sources logic in a separate extension.

I have opened https://github.com/jupyterlab/jupyterlab/issues/11588 to discuss that. For this specific case, I think users want it soon in core and we are already far enough to further progress with this adhoc implementation. I think we also need more toughts and cases to correctly decide on to create the flexible and extensible solution.

Reply:
> 

Were you able to try overriding the handling of that specific request on the kernel side? Instead of forwarding the message to `debugpy`. This might allow running the custom Python code currently sent as a Comm message directly on the kernel instead.

For reference there was something similar in `xeus-python` to be able to open sources: https://github.com/jupyter-xeus/xeus-python/pull/168

Reply:
> Were you able to try overriding the handling of that specific request on the kernel side? Instead of forwarding the message to debugpy. This might allow running the custom Python code currently sent as a Comm message directly on the kernel instead.

I haven't tried to override the Request_Modules, but it sounds something worth to try. Will have a look.

Reply:
>  I have tried to use Requests_Modules https://microsoft.github.io/debug-adapter-protocol/specification#Requests_Modules but it returns only a partial list of the expected modules, not sure why (in my case 13 modules, missing a lot of the existing and loaded modules).

I guess this is because debupy capability `supportsModulesRequest` is [False](https://github.com/microsoft/debugpy/blob/main/src/debugpy/adapter/servers.py#L241), although I don't understand why debugpy returns something instead of throwing an exception.

I think all these requests (ModuleRequest, LoadedSourceRequest) should be implemented on the kernel side. JupyterLab sould only check whether the requets was successful (kernels not supporting these requests should return a response with the success field set to `False`).